lib/concurrent/utility/monotonic_time.rb

```diff
@@ -0,0 +1,147 @@
+module Concurrent
+
+  # Clock that cannot be set and represents monotonic time since
+  # some unspecified starting point.
+  # @!visibility private
+  GLOBAL_MONOTONIC_CLOCK = Class.new {
+
+    if defined?(Process::CLOCK_MONOTONIC)
+      # @!visibility private
+      def get_time
+        Process.clock_gettime(Process::CLOCK_MONOTONIC)
+      end
+    elsif RUBY_PLATFORM == 'java'
+      # @!visibility private
+      def get_time
+        java.lang.System.nanoTime() / 1_000_000_000.0
+      end
+    else
+
+      require 'thread'
+
+      # @!visibility private
+      def initialize
+        @mutex = Mutex.new
+        @correction = 0
+        @last_time = Time.now.to_f
+      end
+
+      # @!visibility private
+      def get_time
+        @mutex.synchronize {
+          @correction ||= 0 # compensating any back time shifts
+          now = Time.now.to_f
+          corrected_now = now + @correction
+          if @last_time < corrected_now
+            @last_time = corrected_now 
+          else
+            @correction += @last_time - corrected_now + 0.000_001
+            @last_time = @correction + now
+          end
+        }
+      end
+    end
+  }.new
```

Comment:
I experimented with a couple different implementations of this and didn't like any of them. The only reason this code exists is because not all runtimes support `Process.clock_gettime(Process::CLOCK_MONOTONIC)`. At ope point I had a set of classes that follow our normal convention but I couldn't think of a single use case for them. I can't think of a single reason a user would need to create an object of this type. All anyone should ever need is `Concurrent.monotonic_time`. The class/object underneath that is just an implementation detail. In an early version I only created `GLOBAL_MONOTONIC_CLOCK` on systems which needed it. I didn't like having a global variable that wasn't available on all platforms so I implemented it universally. But I also wanted to discourage users from creating new instances of the class. A single clock is all that's necessary. So I made it anonymous and hidden from the yardoc.

I agree that it's an odd implementation, but it's also a very odd feature--one that should go away over time as people move to newer versions Ruby.

If the internal inconsistency (with respect to class name and hierarchy) bothers people I don't have a problem with us changing it. In that case I will still want to hide the classes from yardoc. And also add a notice that the class shouldn't be used directly, instead recommending only the use of `Concurrent.monotonic_time`.